src/app/fdctl/run/tiles/fd_dedup.c

```diff
@@ -0,0 +1,186 @@
+
+/* fd_dedup_in_ctx_t is a context object for each in (producer) mcache
+   connected to the dedup tile. */
+
+typedef struct {
+  fd_wksp_t * mem;
+  ulong       chunk0;
+  ulong       wmark;
+} fd_dedup_in_ctx_t;
+
+/* fd_dedup_ctx_t is the context object provided to callbacks from the
+   mux tile, and contains all state needed to progress the tile. */
+
+typedef struct {
+  ulong   tcache_depth;   /* == fd_tcache_depth( tcache ), depth of this dedups's tcache (const) */
+  ulong   tcache_map_cnt; /* == fd_tcache_map_cnt( tcache ), number of slots to use for tcache map (const) */
+  ulong * tcache_sync;    /* == fd_tcache_oldest_laddr( tcache ), local join to the oldest key in the tcache */
+  ulong * tcache_ring;
+  ulong * tcache_map;
+
+  fd_dedup_in_ctx_t in[ 32 ];
+
+  fd_wksp_t * out_mem;
+  ulong       out_chunk0;
+  ulong       out_wmark;
+  ulong       out_chunk;
+} fd_dedup_ctx_t;
+
+FD_FN_CONST static inline ulong
+scratch_align( void ) {
+  return alignof( fd_dedup_ctx_t );
+}
+
+FD_FN_PURE static inline ulong
+scratch_footprint( fd_topo_tile_t * tile ) {
+  (void)tile;
+  ulong scratch_top = 0UL;
+  SCRATCH_ALLOC( alignof( fd_dedup_ctx_t ), sizeof( fd_dedup_ctx_t ) );
+  SCRATCH_ALLOC( fd_tcache_align(), fd_tcache_footprint( tile->dedup.tcache_depth, 0 ) );
+  return fd_ulong_align_up( scratch_top, scratch_align() );
+}
+
+FD_FN_CONST static inline void *
+mux_ctx( void * scratch ) {
+  return (void*)fd_ulong_align_up( (ulong)scratch, alignof( fd_dedup_ctx_t ) );
+}
+
+/* during_frag is called between pairs for sequence number checks, as
+   we are reading incoming frags.  We don't actually need to copy the
+   fragment here, flow control prevents it getting overrun, and
+   downstream consumers could reuse the same chunk and workspace to
+   improve performance.
+
+   The bounds checking and copying here are defensive measures,
+
+    * In a functioning system, the bounds checking should never fail,
+      but we want to prevent an attacker with code execution on a producer
+      tile from trivially being able to jump to a consumer tile with
+      out of bounds chunks.
+
+    * For security reasons, we have chosen to isolate all workspaces from
+      one another, so for example, if the QUIC tile is compromised with
+      RCE, it cannot wait until the sigverify tile has verified a transaction,
+      and then overwrite the transaction while it's being processed by the
+      banking stage. */
+
+static inline void
+during_frag( void * _ctx,
+             ulong  in_idx,
+             ulong  sig,
+             ulong  chunk,
+             ulong  sz,
+             int *  opt_filter ) {
+  (void)sig;
+  (void)opt_filter;
+
+  fd_dedup_ctx_t * ctx = (fd_dedup_ctx_t *)_ctx;
+
+  if( FD_UNLIKELY( chunk<ctx->in[ in_idx ].chunk0 || chunk>ctx->in[ in_idx ].wmark || sz > FD_TPU_DCACHE_MTU ) )
+    FD_LOG_ERR(( "chunk %lu %lu corrupt, not in range [%lu,%lu]", chunk, sz, ctx->in[ in_idx ].chunk0, ctx->in[ in_idx ].wmark ));
```

Comment:
Maybe this is better as a separate PR, but we should consider moving these checks into the mux tile

Reply:
Yeah I already filed an issue for this